liquid-derive/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+#![recursion_limit = "128"]
+extern crate liquid_compiler;
+extern crate liquid_error;
+extern crate proc_macro;
+extern crate proc_macro2;
+extern crate quote;
+extern crate syn;
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::*;
+
+// Probably there is a better approach that doesn't require adding this struct
+// Goal is to be able to use `#` inside a `quote!` that writes all objects inside the `Vec`
+struct ToTokensVec<T: quote::ToTokens>(Vec<T>);
+impl<T: quote::ToTokens> quote::ToTokens for ToTokensVec<T> {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        for token in self.0.iter() {
+            token.to_tokens(tokens);
+        }
+    }
+}
+
+// Generates the statement to construct the field with the given name inside `new(mut args: FilterArguments)`
+fn generate_construct_field(ident: &Ident, is_option: bool) -> proc_macro2::TokenStream {
+    if is_option {
+        quote! {
+            let #ident = args.positional.next();
+        }
+    } else {
+        quote! {
+            let #ident = args.positional.next().ok_or_else(|| liquid_error::Error::with_msg("Required"))?;
+        }
+    }
+}
+
+// Generates the statement to evaluate the field with the given name into a `Value`
+fn generate_evaluate_field(ident: &Ident, is_option: bool) -> proc_macro2::TokenStream {
+    if is_option {
+        quote! {
+            let #ident = match &self.#ident {
+                Some(field) => Some(field.evaluate(context)?.to_owned()),
```

Comment:
So this turns a filter parameter from a variable to a value?  Why do we `to_owned`?  Can't we just pass the returned reference to the filter?

Reply:
Well, in the old code that is what happens, so I thought there wouldn't be much problem with it.
From [`liquid-compiler/src/filter_chain.rs`](https://github.com/cobalt-org/liquid-rust/blob/master/liquid-compiler/src/filter_chain.rs) line 34
```
let arguments: Result<Vec<Value>> = self
            .arguments
            .iter()
            .map(|a| Ok(a.evaluate(context)?.to_owned()))
            .collect();
```

Passing the reference to the filter means turning:
```
struct EvaluatedSliceParameters {
    offset: Value,
    length: Option<Value>,
}
```
into:
```
struct EvaluatedSliceParameters<'a> {
    offset: &'a Value,
    length: Option<&'a Value>,
}
```

Will filters still be able to do everything they need to with a `&Value` instead of a `Value`? Is there any particular reason `Value`s were owned in the old code?

Reply:
They should be able to still work fine.  I've been slowly going through, removing unnecessary allocations.  I missed this one.